crates/sargon/src/profile/mfa/security_structures/matrices/builder/matrix_template.rs

```diff
@@ -0,0 +1,497 @@
+/// materializing a MatrixTemplate into a MatrixOfFactorSourceIds.
+pub(crate) struct FactorSourceIdAssigner {
+    factor_source_ids: Vec<FactorSourceID>,
+    map: IndexMap<FactorSourceTemplate, FactorSourceID>,
+}
+
+impl FactorSourceIdAssigner {
+    fn new(
+        factor_source_ids: impl IntoIterator<Item = FactorSourceID>,
+    ) -> Self {
+        Self {
+            factor_source_ids: factor_source_ids.into_iter().collect_vec(),
+            map: IndexMap::new(),
+        }
+    }
+
+    fn next(
+        &mut self,
+        template: &FactorSourceTemplate,
+    ) -> Result<FactorSourceID, CommonError> {
+        if let Some(existing) = self.map.get(template) {
+            Ok(*existing)
+        } else if let Some(index_of_next) = self
+            .factor_source_ids
+            .iter()
+            .position(|f| f.get_factor_source_kind() == template.kind)
+        {
+            let next = self.factor_source_ids.remove(index_of_next);
+            self.map.insert(template.clone(), next);
+            Ok(next)
+        } else {
+            Err(CommonError::Unknown)
+        }
+    }
+}
+
+impl MatrixTemplate {
+    pub fn materialize(
+        self,
+        factor_source_ids: impl IntoIterator<Item = FactorSource>,
+    ) -> Result<MatrixOfFactorSourceIds, CommonError> {
+        self.materialize_ids(
+            factor_source_ids.into_iter().map(|f| f.factor_source_id()),
+        )
+    }
+
+    /// Tries to materialize a MatrixOfFactorSourceIds from a MatrixTemplate by
+    /// assigning each template with a concrete FactorSourceID using the `factor_source_ids`.`
+    pub fn materialize_ids(
+        self,
+        factor_source_ids: impl IntoIterator<Item = FactorSourceID>,
+    ) -> Result<MatrixOfFactorSourceIds, CommonError> {
+        let number_of_days_until_auto_confirm =
+            self.number_of_days_until_auto_confirm;
+        let mut assigner = FactorSourceIdAssigner::new(factor_source_ids);
+        let primary_role = self.primary_role.assign(&mut assigner)?;
+        let recovery_role = self.recovery_role.assign(&mut assigner)?;
+        let confirmation_role = self.confirmation_role.assign(&mut assigner)?;
+
+        Ok(MatrixOfFactorSourceIds {
+            built: PhantomData,
+            primary_role,
+            recovery_role,
+            confirmation_role,
+            number_of_days_until_auto_confirm,
+        })
+    }
+}
+
+impl MatrixTemplate {
+    fn new(
+        primary_role: PrimaryRoleTemplate,
+        recovery_role: RecoveryRoleTemplate,
+        confirmation_role: ConfirmationRoleTemplate,
+    ) -> Self {
+        Self {
+            built: PhantomData,
+            primary_role,
+            recovery_role,
+            confirmation_role,
+            number_of_days_until_auto_confirm:
+                MatrixOfFactorSourceIds::DEFAULT_NUMBER_OF_DAYS_UNTIL_AUTO_CONFIRM,
+        }
+    }
+
+    pub fn config_11() -> Self {
```

Comment:
Also, would rename them to be `config_1_1` since, for this case in particular (but possible for others in future), we cannot tell if it is `MFA 1.1` or `MFA 11`.

Reply:
FIXED